File: Packages/GymBroCore/Sources/GymBroCore/Services/PersonalRecordService.swift
import Foundation
import SwiftData

public struct PersonalRecord {
    public let exerciseSet: ExerciseSet
    public let recordType: RecordType
    
    public enum RecordType {
        case maxWeight
        case maxReps
        case maxVolume
        case maxE1RM
    }
}

public class PersonalRecordService {
    private let modelContext: ModelContext
    
    public init(modelContext: ModelContext) {
        self.modelContext = modelContext
    }
    
    public func getPersonalRecords(for exercise: Exercise) throws -> [PersonalRecord] {
        let descriptor = FetchDescriptor<ExerciseSet>(
            predicate: #Predicate { set in
                set.exercise?.id == exercise.id && set.setType == .working
            },
            sortBy: [SortDescriptor(\.completedAt, order: .reverse)]
        )
        
        let sets = try modelContext.fetch(descriptor)
        
        guard !sets.isEmpty else { return [] }
        
        var records: [PersonalRecord] = []
        
        let maxWeightSet = sets.max(by: { $0.weightKg < $1.weightKg })
        if let maxWeightSet = maxWeightSet {
            records.append(PersonalRecord(exerciseSet: maxWeightSet, recordType: .maxWeight))
        }
        
        let maxRepsSet = sets.max(by: { $0.reps < $1.reps })
        if let maxRepsSet = maxRepsSet {
            records.append(PersonalRecord(exerciseSet: maxRepsSet, recordType: .maxReps))
        }
        
        let maxVolumeSet = sets.max(by: { $0.volume < $1.volume })
        if let maxVolumeSet = maxVolumeSet {
            records.append(PersonalRecord(exerciseSet: maxVolumeSet, recordType: .maxVolume))
        }
        
        let maxE1RMSet = sets.max(by: { $0.estimatedOneRepMax < $1.estimatedOneRepMax })
        if let maxE1RMSet = maxE1RMSet {
            records.append(PersonalRecord(exerciseSet: maxE1RMSet, recordType: .maxE1RM))
        }
        
        return records
    }
    
    public func isPR(set: ExerciseSet, for recordType: PersonalRecord.RecordType) throws -> Bool {
        guard let exercise = set.exercise, set.setType == .working else { return false }
        
        let descriptor = FetchDescriptor<ExerciseSet>(
            predicate: #Predicate { s in
                s.exercise?.id == exercise.id && 
                s.setType == .working &&
                s.completedAt != nil &&
                s.completedAt! < (set.completedAt ?? Date())
            }
        )
        
        let previousSets = try modelContext.fetch(descriptor)
        
        guard !previousSets.isEmpty else { return true }
        
        switch recordType {
        case .maxWeight:
            return previousSets.allSatisfy { $0.weightKg < set.weightKg }
        case .maxReps:
            return previousSets.allSatisfy { $0.reps < set.reps }
        case .maxVolume:
            return previousSets.allSatisfy { $0.volume < set.volume }
        case .maxE1RM:
            return previousSets.allSatisfy { $0.estimatedOneRepMax < set.estimatedOneRepMax }
        }
    }
    
    public func getRecordTypes(for set: ExerciseSet) throws -> [PersonalRecord.RecordType] {
        var recordTypes: [PersonalRecord.RecordType] = []
        
        for type in [PersonalRecord.RecordType.maxWeight, .maxReps, .maxVolume, .maxE1RM] {
            if try isPR(set: set, for: type) {
                recordTypes.append(type)
            }
        }
        
        return recordTypes
    }
}

File: android/core/src/main/java/com/gymbro/core/service/PersonalRecordService.kt
package com.gymbro.core.service

import com.gymbro.core.database.dao.WorkoutDao
import com.gymbro.core.model.E1RMDataPoint
import com.gymbro.core.model.PersonalRecord
import com.gymbro.core.model.RecordType
import com.gymbro.core.model.WorkoutHistoryItem
import java.time.Instant
import javax.inject.Inject

class PersonalRecordService @Inject constructor(
    private val workoutDao: WorkoutDao,
) {

    /** Brzycki formula: weight * (36 / (37 - reps)) */
    fun calculateE1RM(weight: Double, reps: Int): Double {
        if (weight <= 0 || reps <= 0) return 0.0
        if (reps == 1) return weight
        val denominator = 37.0 - reps
        if (denominator <= 0) return weight * 2.0
        return weight * (36.0 / denominator)
    }

    suspend fun getPersonalRecords(exerciseId: String, exerciseName: String): List<PersonalRecord> {
        val sets = workoutDao.getSetsByExercise(exerciseId)
        if (sets.isEmpty()) return emptyList()

        val records = mutableListOf<PersonalRecord>()

        val maxWeightSet = sets.maxByOrNull { it.weight }
        if (maxWeightSet != null) {
            val previousBest = sets
                .filter { it.completedAt < maxWeightSet.completedAt }
                .maxByOrNull { it.weight }
            records += PersonalRecord(
                exerciseId = exerciseId,
                exerciseName = exerciseName,
                type = RecordType.MAX_WEIGHT,
                value = maxWeightSet.weight,
                date = Instant.ofEpochMilli(maxWeightSet.completedAt),
                previousValue = previousBest?.weight,
            )
        }

        val maxRepsSet = sets.maxByOrNull { it.reps }
        if (maxRepsSet != null) {
            val previousBest = sets
                .filter { it.completedAt < maxRepsSet.completedAt }
                .maxByOrNull { it.reps }
            records += PersonalRecord(
                exerciseId = exerciseId,
                exerciseName = exerciseName,
                type = RecordType.MAX_REPS,
                value = maxRepsSet.reps.toDouble(),
                date = Instant.ofEpochMilli(maxRepsSet.completedAt),
                previousValue = previousBest?.reps?.toDouble(),
            )
        }

        val maxVolumeSet = sets.maxByOrNull { it.weight * it.reps }
        if (maxVolumeSet != null) {
            val volume = maxVolumeSet.weight * maxVolumeSet.reps
            val previousBest = sets
                .filter { it.completedAt < maxVolumeSet.completedAt }
                .maxByOrNull { it.weight * it.reps }
            records += PersonalRecord(
                exerciseId = exerciseId,
                exerciseName = exerciseName,
                type = RecordType.MAX_VOLUME,
                value = volume,
                date = Instant.ofEpochMilli(maxVolumeSet.completedAt),
                previousValue = previousBest?.let { it.weight * it.reps },
            )
        }

        val maxE1RMSet = sets.maxByOrNull { calculateE1RM(it.weight, it.reps) }
        if (maxE1RMSet != null) {
            val e1rm = calculateE1RM(maxE1RMSet.weight, maxE1RMSet.reps)
            val previousBest = sets
                .filter { it.completedAt < maxE1RMSet.completedAt }
                .maxByOrNull { calculateE1RM(it.weight, it.reps) }
            records += PersonalRecord(
                exerciseId = exerciseId,
                exerciseName = exerciseName,
                type = RecordType.MAX_E1RM,
                value = e1rm,
                date = Instant.ofEpochMilli(maxE1RMSet.completedAt),
                previousValue = previousBest?.let { calculateE1RM(it.weight, it.reps) },
            )
        }

        return records
    }

    suspend fun getE1RMHistory(exerciseId: String): List<E1RMDataPoint> {
        val sets = workoutDao.getSetsByExercise(exerciseId)
        if (sets.isEmpty()) return emptyList()

        return sets
            .groupBy { Instant.ofEpochMilli(it.completedAt).epochSecond / 86400 }
            .mapNotNull { (_, daySets) ->
                val best = daySets.maxByOrNull { calculateE1RM(it.weight, it.reps) } ?: return@mapNotNull null
                E1RMDataPoint(
                    date = Instant.ofEpochMilli(best.completedAt),
                    e1rm = calculateE1RM(best.weight, best.reps),
                    weight = best.weight,
                    reps = best.reps,
                )
            }
            .sortedBy { it.date }
    }

    suspend fun getWorkoutHistory(): List<WorkoutHistoryItem> {
        val workouts = workoutDao.getAllCompletedWorkouts()
        return workouts.map { wws ->
            val workingSets = wws.sets.filter { !it.isWarmup }
            val exerciseIds = workingSets.map { it.exerciseId }.distinct()
            WorkoutHistoryItem(
                workoutId = wws.workout.id,
                date = Instant.ofEpochMilli(wws.workout.startedAt),
                exerciseCount = exerciseIds.size,
                totalVolume = workingSets.sumOf { it.weight * it.reps },
                durationSeconds = wws.workout.durationSeconds,
                exerciseNames = emptyList(),
            )
        }
    }

    suspend fun getExerciseIdsWithHistory(): List<String> {
        return workoutDao.getExerciseIdsWithHistory()
    }
}
